### PR TITLE
Propagate swift flavors instead of combining SwiftLibrary & SwiftCompile

### PR DESCRIPTION
**Pull request by nguyentruongtho**
This PR will fix the mistake I made in https://github.com/facebook/buck/pull/831 and fix a problem with flavors propagation mentioned in https://github.com/facebook/buck/pull/843.
A few noticeable changes:
1. When platform flavor is not present, `SwiftLibraryDescription` will create a place holder (`SwiftLibrary`).
2. Create Swift companion target and add it in the extra deps list when necessary, this target will only be created in building target graph phrase.

@ryu2 Please check if I have merge your changes correctly.
@Coneko This is the approach we talked about yesterday, please have a look, love to hear your thought :)

### CODE REVIEW

**Comment by facebook-github-bot-4** [comment]
By analyzing the blame information on this pull request, we identified @ryu2, @Coneko and @beefon to be potential reviewers.

**Comment by Coneko** [comment]
Wonderful! Looks nice!

@facebook-github-bot shipit

**Comment by facebook-github-bot-1** [comment]
Thanks for importing.If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3741867) to review internal test results.

**Comment by facebook-github-bot-8** [comment]
@nguyentruongtho updated the pull request - [view changes](https://github.com/facebook/buck/pull/847/files/9ea02639b2cf7c3c932ddddd034fbbe20ae70911..ea10398b19ba36133c5c8146e319ef84c420067c) - [changes since last import](https://github.com/facebook/buck/pull/847/files/9ea02639b2cf7c3c932ddddd034fbbe20ae70911..ea10398b19ba36133c5c8146e319ef84c420067c)

**Comment by Coneko** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-0** [comment]
Thanks for importing.If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3741867) to review internal test results.

File: test/com/facebook/buck/swift/testdata/swift_on_swift/BUCK.fixture [hunk 1]
```diff
@@ -33,7 +33,7 @@ swift_library(
 apple_binary(
     name = 'ios-parent',
     srcs = [ 'ios-parent/AppDelegate.swift' ],
-    deps = [ ':iosdep1#iphonesimulator-x86_64' ], # TODO remove flavor
```
Comment (nguyentruongtho) [inline test/com/facebook/buck/swift/testdata/swift_on_swift/BUCK.fixture]: err, almost forgot about this test!

**Comment by facebook-github-bot-9** [comment]
@nguyentruongtho updated the pull request - [view changes](https://github.com/facebook/buck/pull/847/files/ea10398b19ba36133c5c8146e319ef84c420067c..9beac7adb61dda12a398f7e8828ee029e6ca9754) - [changes since last import](https://github.com/facebook/buck/pull/847/files/ea10398b19ba36133c5c8146e319ef84c420067c..9beac7adb61dda12a398f7e8828ee029e6ca9754)

**Comment by Coneko** [comment]
`com.facebook.buck.swift.SwiftIOSBundleIntegrationTest swiftWithSwiftDependenciesBuildsSomething` fails: this is with Xcode 7.3.1

```
Undefined symbols for architecture x86_64:
  "type metadata accessor for iosdep1.IosFoo", referenced from:
      ios_parent.AppDelegate.application (__ObjC.UIApplication, didFinishLaunchingWithOptions : [__ObjC.NSObject : Swift.AnyObject]?) -> Swift.Bool in ios_parent.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
BUILD FAILED: //:ios-parent#binary,iphonesimulator-x86_64 failed with exit code 1:
c++ link
stderr: Undefined symbols for architecture x86_64:
  "type metadata accessor for iosdep1.IosFoo", referenced from:
      ios_parent.AppDelegate.application (__ObjC.UIApplication, didFinishLaunchingWithOptions : [__ObjC.NSObject : Swift.AnyObject]?) -> Swift.Bool in ios_parent.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)```

**Comment by Coneko** [comment]
Let me try again with the fix.

**Comment by Coneko** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-7** [comment]
Thanks for importing.If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3741867) to review internal test results.

**Comment by nguyentruongtho** [comment]
I run the test manually and it worked, forgot that @ryu2 has added the test in https://github.com/facebook/buck/pull/842. It should work now.

File: src/com/facebook/buck/swift/SwiftDescriptions.java [hunk 2]
```diff
@@ -53,73 +40,47 @@
...
-        swiftCompanionTarget.getShortName(),
-        swiftSrcs,
-        Optional.of(true),
-        Optional.<Pattern>absent()));
+  static boolean hasSwiftSource(ImmutableSortedSet<SourceWithFlags> srcs) {
+    return FluentIterable.from(srcs)
+        .firstMatch(new Predicate<SourceWithFlags>() {
+          @Override
+          public boolean apply(SourceWithFlags input) {
+            return MorePaths.getFileExtension(Paths.get(input.getSourcePath().toString()))
```
Comment (Coneko) [inline src/com/facebook/buck/swift/SwiftDescriptions.java]: You need to `sourcePathResolver.getAbsolutePath(source.getSourcePath())` here, like above in `filterSwiftSources`.

In fact since it's the same, extract the common logic.

**Comment by Coneko** [comment]
Back to you.

File: src/com/facebook/buck/swift/SwiftDescriptions.java (hunk 2 above)
Comment (nguyentruongtho) [inline src/com/facebook/buck/swift/SwiftDescriptions.java]: I thought about that.
To create a `SourcePathResolver` instance, I will need `BuildRuleResolver`, eventually need `TargetGraph`, and a graph transformer. Problem is that, this function `hasSwiftSource` is being call during `findDepsForTargetFromConstructorArgs` which is in the middle of building a `TargetGraph` instance. 
I could use `TargetGraph.EMPTY` and the default transformer, it should work, but I think it is not necessary, and probably this way works a bit simpler. 

Suggestion?

Comment (nguyentruongtho) [inline src/com/facebook/buck/swift/SwiftDescriptions.java]: nvm, I found a solution for it. just gimme a few mins.

**Comment by facebook-github-bot-3** [comment]
@nguyentruongtho updated the pull request - [view changes](https://github.com/facebook/buck/pull/847/files/9beac7adb61dda12a398f7e8828ee029e6ca9754..9e44e7425a047a8b33ba82e8be1c882df401bb31) - [changes since last import](https://github.com/facebook/buck/pull/847/files/9beac7adb61dda12a398f7e8828ee029e6ca9754..9e44e7425a047a8b33ba82e8be1c882df401bb31)

**Comment by nguyentruongtho** [comment]
Pay off a bit by using a `NoopBuildRule` during creating build rule for apple targets when there is no swift source in the target. However, it looks better to me since most of the logic are now enclosed in `SwiftLibraryDescription`.

**Comment by Coneko** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-6** [comment]
Thanks for importing.If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3741867) to review internal test results.

**Comment by Coneko** [comment]
Travis build is red:

```
FAILURE com.facebook.buck.apple.project_generator.ProjectGeneratorTest stopsLinkingRecursiveDependenciesAtSharedLibraries: //dep:shared#shared,swift-lib
java.lang.NullPointerException: //dep:shared#shared,swift-lib
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:228)
	at com.facebook.buck.testutil.TargetGraphFactory.newInstance(TargetGraphFactory.java:51)
	at com.facebook.buck.apple.project_generator.ProjectGeneratorTest.createProjectGeneratorForCombinedProject(ProjectGeneratorTest.java:4506)
	at com.facebook.buck.apple.project_generator.ProjectGeneratorTest.createProjectGeneratorForCombinedProject(ProjectGeneratorTest.java:4498)
	at com.facebook.buck.apple.project_generator.ProjectGeneratorTest.stopsLinkingRecursiveDependenciesAtSharedLibraries(ProjectGeneratorTest.java:2788)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:239)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at com.facebook.buck.testrunner.SameThreadFailOnTimeout$1.call(SameThreadFailOnTimeout.java:45)
	at com.facebook.buck.testrunner.SameThreadFailOnTimeout$1.call(SameThreadFailOnTimeout.java:41)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

```

**Comment by facebook-github-bot-0** [comment]
@nguyentruongtho updated the pull request - [view changes](https://github.com/facebook/buck/pull/847/files/9e44e7425a047a8b33ba82e8be1c882df401bb31..4615b698165911e93e59613f68016f8fb8b39209) - [changes since last import](https://github.com/facebook/buck/pull/847/files/9e44e7425a047a8b33ba82e8be1c882df401bb31..4615b698165911e93e59613f68016f8fb8b39209)

**Comment by nguyentruongtho** [comment]
Please ignore, I accidentally pushed the changes.

**Comment by facebook-github-bot-0** [comment]
@nguyentruongtho updated the pull request - [view changes](https://github.com/facebook/buck/pull/847/files/4615b698165911e93e59613f68016f8fb8b39209..525f776deff249bda0762689f4c51653d06db137) - [changes since last import](https://github.com/facebook/buck/pull/847/files/9e44e7425a047a8b33ba82e8be1c882df401bb31..525f776deff249bda0762689f4c51653d06db137)

File: src/com/facebook/buck/swift/SwiftDescriptions.java [hunk 3]
```diff
@@ -69,4 +60,27 @@ private SwiftDescriptions() {
     output.enableObjcInterop = Optional.of(true);
   }
 
+  static boolean hasSwiftSource(ImmutableSortedSet<SourceWithFlags> srcs) {
+    return Iterables.any(srcs, new Predicate<SourceWithFlags>() {
+      @Override
+      public boolean apply(SourceWithFlags input) {
+        SourcePath sourcePath = input.getSourcePath();
+        return (sourcePath instanceof PathSourcePath) &&
```
Comment (nguyentruongtho) [inline src/com/facebook/buck/swift/SwiftDescriptions.java]: @Coneko 
Since we couldn't obtain a proper instance of `SourcePathResolver` at this point (explained here https://github.com/facebook/buck/pull/847#discussion_r75711254) and my last attempt (by always add a swift companion target for the apple target) didn't work well because it messed up with the `TargetGraph`, we can't share the logic with `filterSwiftSources`. 
If we look at `SourcePathTypeCoercer`, there are only 2 possible types of a `SourcePath` in here, `BuildTargetSourcePath` and `PathSourcePath`, but `BuildTargetSourcePath` is not for element in `srcs` so we can cast sourcePath to `PathSourcePath` and check for its extension.

Comment (Coneko) [inline src/com/facebook/buck/swift/SwiftDescriptions.java]: `srcs` can contain `BuildTargetSourcePath`: it can refer to files that are outputs of other rules, for example when using `genrule` to generate sources.

Comment (Coneko) [inline src/com/facebook/buck/swift/SwiftDescriptions.java]: I think the approach of always having a fake swift placeholder rule, and just having it do nothing if there are no swift sources, which you used in the previous commit was good.

Comment (nguyentruongtho) [inline src/com/facebook/buck/swift/SwiftDescriptions.java]: The previous approach has 2 issues:
1. Since there are extra `swift-lib` targets even when there is no swift source, project generation task will have to filter them out, this is consider overhead. Indeed, project generation has nothing to do with the companion target, we should better off generate xcode project without creating the companion target.
2. There is the case that `AppleTestDescription` delegate its `findDepsForTargetFromConstructorArgs` to corresponding `AppleLibraryDescription`, here is where the mess gets bigger, since we have a companion target for `apple_test` as well, this is make the project generation way more complicated.

Since I haven't started with project generation, I would not touch before I have a clear picture of this part. I will update PR to cover the case of `BuildTargetSourcePath` and will update this when we start to work with project generation.
The change in code would not be big though, sound good to you @Coneko ?

Comment (nguyentruongtho) [inline src/com/facebook/buck/swift/SwiftDescriptions.java]: Just looked at the how `BuildTargetSourcePath` can be included in `srcs` (like the one in `genrule`), we don't know that path without a `BuildRuleResolver` (https://github.com/facebook/buck/blob/405d49e6685071f12884998c0b3066dcdcc67193/src/com/facebook/buck/rules/SourcePathResolver.java#L191-L191). But the case of using genrule to add swift source file is quite trivial at this moment, we can just ignore it for now. I'll revisit this again when working on project generation.

Comment (Coneko) [inline src/com/facebook/buck/swift/SwiftDescriptions.java]: Good points about the project generation. `halide_library` used to implement `findDepsForTargetFromConstructorArgs` in a similar way but then we removed it because it was adding complexity for nothing.

The dependency on the placeholder swift library can be added in `createBuildRule`, `findDepsForTargetFromConstructorArgs` just makes it easier to add, but the additional complexity down the line is not worth it.

Can you change `AppleLibraryDescription.createBuildRule` and `AppleBinaryDescription.createBuildRule` to do that and remove the changes to `findDepsForTargetFromConstructorArgs`?

Comment (Coneko) [inline src/com/facebook/buck/swift/SwiftDescriptions.java]: Note you can use `requireRule` and the `swift-library` flavour to get the rule you need, which will go through the same code path as returning the target with that flavour from `findDepsForTargetFromConstructorArgs`: the only thing that changes is where you add the rule as a dependency, not how you get the rule.

Comment (nguyentruongtho) [inline src/com/facebook/buck/swift/SwiftDescriptions.java]: That's a great idea, actually, it is quite similar to my original approach (the one with modification done in `DefaultTargetNodeToBuildRuleTransformer`) in the sense of not touching `TargetGraph` but only modify the `ActionGraph`. 
Changes is coming.

**Comment by facebook-github-bot-2** [comment]
@nguyentruongtho updated the pull request - [view changes](https://github.com/facebook/buck/pull/847/files/525f776deff249bda0762689f4c51653d06db137..5c3b3f540ec6937b7aa74ce657057e1e7519c5cc) - [changes since last import](https://github.com/facebook/buck/pull/847/files/9e44e7425a047a8b33ba82e8be1c882df401bb31..5c3b3f540ec6937b7aa74ce657057e1e7519c5cc)

File: src/com/facebook/buck/swift/SwiftLibraryDescription.java [hunk 4]
```diff
@@ -90,45 +118,103 @@ public boolean hasFlavors(ImmutableSet<Flavor> flavors) {
...
+        new SourcePathResolver(resolver),
+        delegateArgs,
+        args,
+        buildTarget);
+    if (delegateArgs.srcs.isPresent() && !delegateArgs.srcs.get().isEmpty()) {
+      return Optional.of(
+          resolver.addToIndex(
+              createBuildRule(targetGraph, params, resolver, delegateArgs)));
+    } else {
+      return Optional.absent();
```
Comment (Coneko) [inline src/com/facebook/buck/swift/SwiftLibraryDescription.java]: I don't think we should ever hit this code path.

**Comment by Coneko** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-6** [comment]
Thanks for importing.If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3741867) to review internal test results.

File: src/com/facebook/buck/swift/SwiftLibraryDescription.java (hunk 4 above)
Comment (nguyentruongtho) [inline src/com/facebook/buck/swift/SwiftLibraryDescription.java]: You are absolutely right.
